### ISSUE TITLE
FEATURE REQUEST: Ability to listen to particular events for overrides/cancelling EC timer.

### ISSUE THREAD

**Issue by shbatm** [danobot/entity-controller#106]
**Is your feature request related to a problem? Please describe.**
I use Insteon switches for most of my lighting, with an ISY994 controller feeding back to HA.  Right now, if I use a light controlled by a Insteon switch as a controlled entity with this component, there's no way to cancel the timer with the physical switch without dimming/brightening the light (if its a dimmer) or turning off and back on if its a on/off switch.  

For example, EC turns the light on based on a motion sensor, and Home Assistant sees the state of the light as ON from the ISY controller. When you press the physical switch ON again while the EC timer is running, the light's state doesn't change in HA, it's still ON; however, there is a  "isy994_control" event is sent which has data that says "someone turned the switch on". I'd like to treat this the same as a dim/brighten/state change in EC.

**Describe the solution you'd like**
This project would be perfect for my motion lights if there was an ability to listen to a custom HA event when monitoring to see if a device turns on.  In my example, I would like to listen for the `isy994_control` events with a given entity_id/command and use that to cancel/override the EC timer (to basically tell EC the same as "this entities brightness/color/state has changed".

**Describe alternatives you've considered**
n/a

**Additional context**
I'll take a look at the code in more detail and see if this is something I can implement and trial, but I wanted to share my interest.  Please let me know if the above doesn't make sense.

**Comment by danobot** [comment]
It should be something like a configurable list of enitty attributes that will be ignored on state updates.
e.g.

  mtn_office:
    sensor: binary_sensor.office_motion
    trigger_on_activate: light.office_led
    delay: 120
    state_attributes_ignore:
        - brightness
        - color
Don't make it specific to brightness and color, it should be configurable for any type of entity :) Great idea actually

**Comment by danobot** [comment]
I think the key ability you are looking for is to control the EC entity itself, whether that is by events or services. Support for service calls has been on my backlog for a long time, but I do not have any use cases for it myself. What interactions would you find useful? 

We would need to support event alternatives for differnet entity types. Users may want to listen to events to replace sensor entities and override entities (maybe more).
Note that this implementation is not possible for state entities, because state entities are stateful (have a certain state for a period of time), whereas events are instantaneous.

I would like to reuse as much of the current configuration logic as possible. One way to start supporting entities would be with a configuration as follows:

```
entity_controller:
  event_test_ec:
    sensors: 
      - input_boolean.motion_sensor
      - '#isy994_control'
    entities:
      - light.table_lamp

```
The above EC instance listens to the state of a conventional sensor entity and also listens to the `isy994_control` event. This is configured using a special syntax within the `sensors` configuration attribute. Sensor entities are configured in the `config_sensor_entities`, which needs to be changed to recognise the "event syntax" and register even listeners on `self.hass.bus` to be notified when those events occur.
Write a generic method that will parse the config list and seperate is out into two lists - one containing all conventional entities, and another list containing the event names. Take a look at the `add` method which is being used in all `config` methods. Maybe something cna be done in there.

Thats just to get you started, let me know what you think. There may be a better way to achieve what you need. I do want to avoid creating new config entires such as `sensor_events:` when the exisitng `sensors` entry can be reused nicely.

Events are instantaneous, which means that the EC cannot have the `sensor_type: duration` configuration entry if any entity list cnotains an event entry.

**Comment by shbatm** [comment]
Apologies, I missed your response over the holidays and I haven't had the time to really start fiddling with this myself yet.

> We would need to support event alternatives for differnet entity types. Users may want to listen to events to replace sensor entities and override entities (maybe more).

This is exactly what I'm after. I have 2 usage cases I'll go into more detail below, but yes, I'd like events to be available as alternatives for the different types: sensors, overrides, and control entities.

> Events are instantaneous, which means that the EC cannot have the sensor_type: duration configuration entry if any entity list cnotains an event entry.

Agree. For sensors, you could only have an instantaneous trigger.

> I would like to reuse as much of the current configuration logic as possible.

My thought would be to use a special syntax under each list item. If we use templates for events (see next comment below) you could just check for the "`{`" at the start of the string.

```
entity_controller:
  event_test_ec:
    sensors: 
      - input_boolean.motion_sensor
      - "{% value_json['event_type'] == 'isy944_control' and value_json['data']['entity_id'] == 'light.table_lamp' and value_json['data']['control'] == 'DON' %}
    entities:
      - light.table_lamp
```

> Do you depend on event data at all?

Yes. The events I'm considering rely on event data which contains the entity ID and the specific command called. My thoughts on this would be to use a `async_track_template` call for the events instead of `async_track_state_change`.  See [here](https://dev-docs.home-assistant.io/en/master/api/helpers.html#module-homeassistant.helpers.event) and [here](https://github.com/home-assistant/home-assistant/blob/fbc53a886f9db887b10b03318e9de568160b1839/homeassistant/helpers/event.py#L115) for the built-in functions; looking at the code, I'd probably have to duplicate the helper function because it's a one-shot right now, not repeating. Unless you wanted to re-initialize the listener every time it fires.


## Use Case 1:

1. Upstairs Hallway Light is controlled by 2 Insteon Switches and a motion detector.
2. Switch 1 actually has the load attached to it.
3. Switch 2 is a "virtual" 3-way switch. INSTEON & ISY take care of keeping the status in sync with Switch 1.
4. Switch 1 shows up in HASS as ON when the light is ON regardless of how it's turned on (it is both the control entity and the state entity).
5. Turning either of the physical switches ON (hitting the button) will not report a new ON state in HASS (it's already on), but it *will* fire an `isy994_control` event with a payload like below:

```json
{
    "event_type": "isy994_control",
    "data": {
        "entity_id": "light.switch1",
        "control": "DON",
        "value": "255"
    },
    "origin": "LOCAL",
    "time_fired": "2020-01-08T11:00:34.906146+00:00",
    "context": {
        "id": "f05f39d3fa4544f89111f5fab2df63b4",
        "parent_id": null,
        "user_id": null
    }
}
```
6. I would like an ON event fired by either switch 1 or switch 2, when the entity_controller is ACTIVE to keep the light on (I think this is the OVERRIDE state, it is Scenario 3.2 in your original motion light context in the README).  In real life: motion sensor turns light on and someone hits the ON on the switch to keep the light on.
7. I would like an OFF event for either switch, fired in the last for example 30 seconds, to override the controller (equivalent to an overrides entity), and keep the light OFF even if the motion sensor is triggered (in real life: hit the off on the switch before the motion sensor is triggered to keep the lights from turning on while you walk up/down the stairs--*sneak mode*).
    - This may require another timer, or using the block_timeout property. This case looks to be an 'inverted' `block_timeout` (e.g. keep off for x seconds, instead of keep on).

## Use Case 2:

While thinking through the above, I thought of another possible use case using Alexa. 

1. You could use Alexa events to keep the controlled light on, instead of hitting the switch. For example, saying "Alexa, turn on upstairs light" while the controller is ACTIVE becomes the same as scenario 3.2 and keeps the light on.
2. You could use the Alexa event as a sensor. For example, tell Alexa to "turn on the patio lights", and that activates the controller to turn them off after X amount of time.

The Alexa event looks like:

```json
{
  "event_type": "alexa_smart_home",
  "data": {
    "request": {
      "namespace": "Alexa.PowerController",
      "name": "TurnOn",
      "entity_id": "light.office_lamps"
    },
    "response": {
      "namespace": "Alexa",
      "name": "Response"
    }
  },
  "origin": "LOCAL",
  "time_fired": "2020-01-08T09:16:38.175289+00:00",
  "context": {
    "id": "26655eae5a9549ffa7d2ecbd0e0d4d16",
    "parent_id": null,
    "user_id": "3133d2d7c80a476888e7a6c8e3790bc8"
  }
}
```

## Comments:
1. I recognize I can do most of what I'm looking for above by adding some extra `input_booleans` that are controlled by external automatons looking at these events; but if this is something that can all be baked into this component, I think it makes for a cleaner overall approach.

**Comment by danobot** [comment]
Feel free to submit a PR for this and I'll take a look.
Why do you need templates for events? Wouldn't the event name suffice?

**Comment by shbatm** [comment]
> Feel free to submit a PR for this and I'll take a look.

OK. I hope to be able to work on in over the next couple weeks.

> Why do you need templates for events? Wouldn't the event name suffice?

The event name doesn't have any info about the calling entity or the specific command. `isy994_control` for example fires for EVERY light, switch, thermostat change, sensor detection, etc. in my system--it'd be the MQTT equivalent of listening to the `'#'` topic.

**Comment by danobot** [comment]
This soudns like two feature requests. One is for template support already raised in #8 but applied to all confguratino (not just service_params). The second one if the ability to listen to events with additinoal functinoality for filtering events based on event data.
This is a lot of work by the looks of it.

**Comment by shbatm** [comment]
I'm willing to take a stab at it--this is necessary functionality for me, so it's either write my own NR flow/AD App or make the necessary changes here. I'd like to just build on what you have already.

**Comment by shbatm** [comment]
The more I started playing with this the more I think #125 needs to be implemented and would solve most of my issues. Assuming you had full control of the entity_controller from an external automation or Node-RED, then I could do the advanced controls from events without having to spend too much time re-inventing the wheel with event listening.

I pushed #133 as a basic implementation of services. It still needs some work (see the `# TODO:`s) to actually do anything useful with the controller, but it will give you a good jump on implementing for #125. I will work on expanding it when I can.
